### ISSUE TITLE
Option to limit sync to local addresses

### ISSUE THREAD

**Issue by drws** [syncthing/syncthing#8078]
There are scenarios when global discovery is needed or preferred between two hosts in the same LAN or local discovery is not available. From my experience with global discovery in such cases, Syncthing sometimes discovers _correct_ LAN addresses and in other cases global (relayed) addresses are found. I'm proposing to add an (advanced?) option to limit sync to local addresses only. By setting the option, ST would sync only with hosts that are discovered in the same LAN even when global discovery is used.

**Comment by acolomb** [comment]
Would the [allowedNetworks setting](https://docs.syncthing.net/advanced/device-allowednetworks.html) suit your needs?

**Comment by AudriusButkevicius** [comment]
Yes, that's already done via that option.

**Comment by drws** [comment]
@acolomb Thank you for information. The `allowedNetworks` option is related, but unfortunately **no, it does not** provide what is needed here. That would be something along `allowedNetworks=direct`, where ST allowed all IP ranges up to 1 hop away (directly-attached networks) given the current and ever-changing IP configuration. `allowedNetworks` currently unfortunately provides only static (hardcoded) network list.

One might suggest setting `allowedNetworks=10.0.0.0/8,172.16.0.0/12,192.168.0.0/16` in order to achieve what I'm describing, but this is still not nearly a solution. First of all, such static list allows all private IPs, not only the ones used in directly-attached V(LAN) and on the other hand it prevents ST from working in a network that uses any other IP ranges. It is also not a viable solution for IPv6 and so on and so on...

@AudriusButkevicius It is unfortunately not... Please revert the premature closing since it appears I cannot :)

**Comment by AudriusButkevicius** [comment]
Right, so I understand your point better, but I'm not convinced that a feature like this would add much and whether would have any other users other than you.

The only difference between what you suggested and allowed networks is that it becomes relative to the interfaces the device has.

There are RFCs that declare what CIDRs are to be used for LANs, so I think its fair to assume that if you use the CIDRs from the RFC that limits the connections to local network.

For ipv6, there is no "local network". There are link local addresses, which also have a well known CIDR.

Personally, I think it should stay closed, but as always, other maintainers might disagree and reopen this.

**Comment by calmh** [comment]
The described use case is to use global discovery but be guaranteed to not use random relays or outside-detected NAT addresses. Setting the allowed networks list to your set of possible LAN addresses seems to cover this use case.

**Comment by acolomb** [comment]
As described previously, the implementation to determine whether some address is "one hop away" is *far* from trivial and probably all but consistent between supported operating systems. Definitely the effort is much more than the possible benefit here.

You could use the REST API to manipulate the `allowedNetworks` setting whenever a local network interface address changes. But I don't see it as a Syncthing core functionality either.

**Comment by drws** [comment]
> There are RFCs that declare what CIDRs are to be used for LANs, so I think its fair to assume that if you use the CIDRs from the RFC that limits the connections to local network.

There is a possibility that the two hosts are in a same public subnet, which cannot be cleanly solved with current implementation of `allowedNetworks`. Also, if all private IPs are allowed, traffic can still leak to routable local addresses (such as in the case of "modern" NATted mobile networks). Preserving mobile data also poses a nice example that would be solved with the proposed feature.

> The described use case is to use global discovery but be guaranteed to not use random relays or outside-detected NAT addresses.

Thank you for a spot-on recap of the general idea!

> Setting the allowed networks list to your set of possible LAN addresses seems to cover this use case.

The problem is that the possible addresses change during every network change. While I agree it might not be a very common case, locally-syncing two personal devices while travelling and trying to not burn capped internet data should be a familiar scenario. Also from a security standpoint, there is no need for data to sync over the internet if the other host is in the very same network. Now that I think about it, it also increases ST protocol efficiency in these specific cases. I see the mood is not right, but some other time we might be discussing employing such mechanism by default, because why exactly not?

> the implementation to determine whether some address is "one hop away" is _far_ from trivial and probably all but consistent between supported operating systems.

ST probably already knows which interfaces it listens on? One possible solution would be to calculate relevant subnets from configured IPs and masks. Another is to check the OS's routing table which should have a predictable metric for directly-attached networks.

> You could use the REST API to manipulate the `allowedNetworks` setting whenever a local network interface address changes. But I don't see it as a Syncthing core functionality either.

Yes and it would probably take me no more effort than cooperating in this discussion, but I believe it is worth trying and it would just a bit more improve an already awesomely-hackable piece of software. There are such details settable along (numerous!) ST's advanced options there that this one wouldn't even stand out!

As it could be introduced as an extension to `allowedNetworks`, it wouldn't even take up any new space. It would be complementary to current implementation, for example: `allowedNetworks=direct,192.168.0.0/16`. However, I do agree that the question how to properly and reliably detect directly-attached subnets needs to be answered first.

**Comment by acolomb** [comment]
> ST probably already knows which interfaces it listens on? One possible solution would be to calculate relevant subnets from configured IPs and masks. Another is to check the OS's routing table which should have a predictable metric for directly-attached networks.

> There are such details settable along (numerous!) ST's advanced options there that this one wouldn't even stand out!

I suggest you start digging into how this would actually be solved in code.  Maybe my estimation is wrong, but I still think you are oversimplifying the issue.  It's not about screen space or number of options, but the hard problem is guessing what is "local", which effectively requires knowing how the operating system does network routing.  Put firewall rules into the mix and all that "knowledge" may be completely wrong.

If only local sync is desired, users can either set `allowedNetworks` or disable global discovery.  After all, local sync should work using only local discovery.  If it doesn't work on your network, that is already an uncommon special case.

Another solution that might work for your requirements would be to restrict global routing for Syncthing's traffic in your local firewall.  On that level, it's probably easier to distinguish "local" from "external" IPs.

Putting this functionality into Syncthing core was rejected by several maintainers.  If you want to discuss further, please move it to the [forum](https://forum.syncthing.net/).  If you come up with working and portable code to check whether an IP address is considered local, I'm all ears, but still the forum is the right place for it.

**Comment by bt90** [comment]
> ST probably already knows which interfaces it listens on? One possible solution would be to calculate relevant subnets from configured IPs and masks. Another is to check the OS's routing table which should have a predictable metric for directly-attached networks.

That doesn't really work for all usecases. As an example:

* `192.168.178.0/24` LAN
* `10.0.0.0/8` DMZ

I can handle routing on the gateway without clients of both subnets knowing anything about each other. The `allowedNetworks` option covers this just fine.

The IPv6 argument is also not that strong because we're talking about local networks where we can rely IPv4 addressing. It's also possible to additionally use *site local* IPv6 addresses if somebody wants to and specify those in the `allowedNetworks`.

tl;dr networking is complex and clients might not even have the necessarry information to decide if something should be considered local or not.

**Comment by drws** [comment]
> I suggest you start digging into how this would actually be solved in code.

I was unnecessarily complicating things before with routing table and subnet calculation. Checking the discovered address's presence in ARP table and possibly generating a request if it's not to double-check should be a better solution.

> After all, local sync should work using only local discovery. If it doesn't work on your network, that is already an uncommon special case.

Local discovery needs a server to run on a designated port on all but one device (N-1) in a cluster of (N) devices that need to be synced locally in order to work in every possible device combination or am I missing something? Some devices might currently not be with you, some might have restrictive security policies and can only run clients, some might be simply turned off and there always needs to be at least one able to run the local disco.

> Another solution that might work for your requirements would be to restrict global routing for Syncthing's traffic in your local firewall.

While external and very similar to `allowedNetworks` filter, that is another possible solution that I'll keep in mind. In this discussion let's just finish dissecting if there is some idea in all of this that could be beneficial for Syncthing. I see this discussion as very near its end and see no need to pursue additional one in the forum. For the sake of civil conversation and completeness I would only like to find answers to remaining questions that are related to this particular issue (see below).

> Putting this functionality into Syncthing core was rejected by several maintainers.

The idea of hacky local-address detection was, with a revision option if a better solution is found, about which you are also talking yourself. I always agreed that a versatile detection solution is a non-trivial prerequisite and I never wanted half-working solutions. Please help with the remaining questions below so we finish this ASAP. The questions might seem long since they include explanations for possible newcomers, but you can answer 3 out of 4 with a simple _yes_/_no_.

**The remaining questions, directly related to the issue, are:**

1. Simple and reliable local-addresses detection is a prerequisite for even considering this feature request. The current idea is looking up the globally-found host IP in the ARP table. The problem is that there is no guarantee host is in the table even if on the same network, so probably an ARP request with the host's IP needs to be sent out before. The question for the devs is their **opinion on using the ARP** table/requests for local-address detection and somewhat obviously if there is a better way to do that?

2. From my experience I can confirm that there are cases where data is relayed globally instead of trying just a little bit harder to discover the possible local route. This is generating unnecessary internet traffic in both uplink and downlink direction, taking up (precious?) relaying resouces and can decrease sync speed considerably. Again, that is not only a theory, I've seen it while having global discovery enabled. The question for the devs here is simple: Should we try to find better solution proposals or **is a premature global address detection not considered a problem**?

3. In the "premature global address detection" case, two hosts in the same network find a global (possibly relayed) route between them. When that happens, there is a question on how a possible subsequent local-address detection is handled. **Does ST switch to a local route** after sync already started through a global one?

4. While the other questions are about performance, this one is about a **security** perspective of the matter. Limiting the data to not be sent outside the local network might be desirable by some users. Syncthing seems like a very privacy-aware application and if some users do not want their data to traverse the Internet (encrypted or not) while still having global discovery enabled (possibly because of some non-related folder needing it), **is that within the scope of the project?**

**Comment by imsodin** [comment]
This discussion is a prime example of: Requires further discussion for clarification and narrowing before it becomes actionable. And those discussions should happen on the forum. Or differently put: I agree with the general sentiments before (this is out of scope), and then I also see some points which are interesting or require clarifications, thus there's merit in continuing. However on the forum: Please open a topic on https://forum.syncthing.net - thanks!
